WordPress/src/main/java/org/wordpress/android/ui/media/services/MediaUploadService.java

```diff
@@ -98,75 +106,105 @@ public int onStartCommand(Intent intent, int flags, int startId) {
-        if (mCompletedItems == null) {
-            mCompletedItems = new ArrayList<>();
-        }
-        return mCompletedItems;
-    }
-
     private void handleOnMediaUploadedSuccess(@NonNull OnMediaUploaded event) {
         if (event.canceled) {
             // Upload canceled
             AppLog.i(AppLog.T.MEDIA, "Upload successfully canceled.");
-            completeCurrentUpload();
+            completeUploadWithId(event.media.getId());
             uploadNextInQueue();
         } else if (event.completed) {
             // Upload completed
             AppLog.i(AppLog.T.MEDIA, "Upload completed - localId=" + event.media.getId() + " title=" + event.media.getTitle());
-            mCurrentUpload.setMediaId(event.media.getMediaId());
-            completeCurrentUpload();
+            // here we need to edit the corresponding post
+            updatePostWithMediaUrl(event.media);
+            completeUploadWithId(event.media.getId());
             uploadNextInQueue();
+            stopServiceIfUploadsComplete();
         } else {
             // Upload Progress
-            // TODO check if we need to broadcast event.media, event.progress or we're just fine with
+            // TODO check if we need to re-broadcast event.media, event.progress or we're just fine with
             // listening to  event.media, event.progress
+            AppLog.d(AppLog.T.MEDIA, event.media.getId() + " - progressing " + event.progress);
         }
     }
 
     private void handleOnMediaUploadedError(@NonNull OnMediaUploaded event) {
         AppLog.w(AppLog.T.MEDIA, "Error uploading media: " + event.error.message);
         // TODO: Don't update the state here, it needs to be done in FluxC
-        mCurrentUpload.setUploadState(UploadState.FAILED.name());
-        mDispatcher.dispatch(MediaActionBuilder.newUpdateMediaAction(mCurrentUpload));
-        completeCurrentUpload();
-        // TODO: check whether we need to broadcast the error or maybe it is enough to register for FluxC events
-        // event.media, event.error
+        MediaModel media = getMediaFromQueueById(event.media.getId());
+        if (media != null) {
+            media.setUploadState(UploadState.FAILED.name());
+            mDispatcher.dispatch(MediaActionBuilder.newUpdateMediaAction(media));
+        }
+        completeUploadWithId(event.media.getId());
         uploadNextInQueue();
     }
 
-    private void uploadNextInQueue() {
-        // waiting for response to current upload request
-        if (mCurrentUpload != null) {
-            AppLog.i(AppLog.T.MEDIA, "Ignoring request to uploadNextInQueue, only one media item can be uploaded at a time.");
-            return;
+    private synchronized void updatePostWithMediaUrl(MediaModel media){
+        if (media != null) {
+            PostModel post = mPostStore.getPostByLocalPostId(media.getPostId());
+            if (post != null) {
+                String content = post.getContent();
+                if (content != null) {
```

Comment:
Following up on this - https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/375 adds annotations to `PostModel` getters so cases like these are easier to spot,